master/internal/core_checkpoint.go

```diff
@@ -0,0 +1,397 @@
+package internal
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/google/uuid"
+	"github.com/labstack/echo/v4"
+	"github.com/pkg/errors"
+
+	"github.com/determined-ai/determined/master/internal/api"
+	"github.com/determined-ai/determined/master/pkg/ptrs"
+	"github.com/determined-ai/determined/master/pkg/schemas/expconf"
+)
+
+type archiveWriter interface {
+	WriteFileHeader(fname string, size int64) error
+	Write(b []byte) (int, error)
+}
+
+type tarArchiveWriter struct {
+	tw *tar.Writer
+}
+
+func (aw *tarArchiveWriter) WriteFileHeader(fname string, size int64) error {
+	hdr := tar.Header{
+		Name: fname,
+		Mode: 0666,
+		Size: size,
+	}
+	if strings.HasSuffix(fname, "/") {
+		// This a directory
+		hdr.Mode = 0777
+	}
+	return aw.tw.WriteHeader(&hdr)
+}
+
+func (aw *tarArchiveWriter) Write(p []byte) (int, error) {
+	return aw.tw.Write(p)
+}
+
+type zipArchiveWriter struct {
+	zw        *zip.Writer
+	zwContent io.Writer
+}
+
+func (aw *zipArchiveWriter) WriteFileHeader(fname string, size int64) error {
+	// Zip by default sets mode 0666 and 077 for files and folders respectively
+	zwc, err := aw.zw.Create(fname)
+	if err != nil {
+		return err
+	}
+	aw.zwContent = zwc
+	return nil
+}
+
+func (aw *zipArchiveWriter) Write(p []byte) (int, error) {
+	var w io.Writer
+	if aw.zwContent == nil {
+		return 0, nil
+	}
+	w = aw.zwContent
+	return w.Write(p)
+}
+
+type delayWriter struct {
+	delayBytes int
+	buf        []byte
+	next       io.Writer
+}
+
+func (w *delayWriter) Write(p []byte) (int, error) {
+	if w.buf == nil {
+		return w.next.Write(p)
+	}
+
+	w.buf = append(w.buf, p...)
+	if len(w.buf) < w.delayBytes {
+		return len(p), nil
+	}
+
+	n, err := w.next.Write(w.buf)
+	w.buf = nil
+	return n, err
```

Comment:
This implementation violates the io.Writer interface, since you can't return `n > len(p)`, but if you have a bunch of buffered bytes, the last `return n, err` will always violate that.

Also, this introduces an unnecessary data copy right when the buf exceeds the limit.

I think a better approach would be like:
```go
if w.buf != nil {
  if len(w.buf) + len(p) < w.delayBytes {
    // not enough bytes yet, just wait
    w.buf = append(w.buf, p...)
    return len(p), nil
  }
  // enough bytes, flush old data before writing current data
  _, err := w.next.Write(w.buf)
  w.buf = nil
  if err != nil {
    return 0, err
  }
}
return w.next.Write(p)
```